repoarchivetool/data_retrieval.py

```diff
@@ -0,0 +1,248 @@
+import requests
+import datetime
+
+from github_api_toolkit import github_interface
+
+def get_organisation_repos(org: str, date: str, repo_type: str, gh: github_interface) -> str | list:
+    """ 
+        Gets all repositories which fit the given parameters.
+
+        ==========
+
+        Makes a test call to the API.
+        If successful, get the total number of repository pages (2 repositories per page)
+        Convert the given string, date, to a date object.
+        Calculate the midpoint of the repositories, which denotes where the given archive
+        date is.
+        Any repositories to the left of the midpoint page are newer than the given date,
+        and cannot be archived.
+        Any repositories to the right of the midpoint page are newer than the given date,
+        and can be archived.
+        For each repository between the midpoint and last page of repositories,
+        add the repository to reposToArchive but only if it hasn't already been archived.
+        Return reposToArchive.
+
+        Args:
+            org (str): The name of the organisation whose repositories are to be returned.
+            date (str): The date which repositories that have been committed prior to will be archived.
+            repo_type (str): The type of repository to be returned (public, private, internal or all).
+            gh (api_controller): An instance of the APIHandler class to interact with the Github API.
+
+        Returns:
+            str: An error message.
+            or
+            list: A list of dictionaries containing information about the repositories collected from 
+            the Github API.
+    """
+
+    
+    def archive_flag(repo_url: str, comp_date: date) -> bool | str:
```

Comment:
Does this need to be a nested function? Can we move it out from get_organisation_repos as I find it harder to read the intent of that function.